drawer api refactor

hellochar opened the pull request:
Create a SVGDrawer and a CanvasDrawer, and have Drawer just proxy between them.

hellochar committed:
refactor drawers into SVGDrawer and CanvasDrawer

hellochar committed:
fix tests

themadcreator commented on src/drawers/drawer.ts:
```diff
@@ -5,238 +5,130 @@
+  public remove() {
+    if (this._svgRootG != null) {
+      this._svgRootG.remove();
+    }
```
should this remove that canvas as well?

themadcreator replied:
Also another opportunity to delegate to a common interface between the drawer types

```typescript
interface IDrawer {
    remove: () => void;
    ...
}
```

themadcreator commented on src/drawers/drawer.ts:
```diff
@@ -5,238 +5,130 @@
-        Utils.Window.setTimeout(() => this._drawStep(drawStep), delay);
-        delay += drawStep.animator.totalTime(data.length);
-      });
+    if (this._svgRootG != null) {
```
maybe this should null check the _svgDrawer and the one below can check _canvasDraw

themadcreator commented on src/drawers/drawer.ts:
```diff
@@ -5,238 +5,130 @@
-  public remove() {
-    if (this.renderArea() != null) {
-      this.renderArea().remove();
+  public useCanvas(canvas: d3.Selection<HTMLCanvasElement, any, any, any>) {
```
given this mutual exclusion, maybe the HTMLElement (svg or canvas) should be managed by the respective drawer object.

for example, below, there is the .selection() method but that only applies to svg drawers that have a svg comeponent.

themadcreator commented on src/drawers/drawer.ts:
```diff
@@ -5,238 +5,130 @@
-      }
-      this._cachedSelectionValid = true;
+  public selectionForIndex(index: number): SimpleSelection<any> {
+    if (this._svgRootG != null) {
```
more checks that could be put on the SVGDrawer interface only

themadcreator commented on src/drawers/lineDrawer.ts:
```diff
@@ -5,56 +5,40 @@
+    const d3Line = d3LineFactory();
     const resolvedAttrs = Object.keys(attrToAppliedProjector).reduce((obj, attrName) => {
-      obj[attrName] = attrToAppliedProjector[attrName](data, 0);
+      if (attrName !== "d") {
```
heh, i bet this saves some time!

leebyp replied:
related to the other thread of parsing out some of the "stylistic attribs", some differentiation of which attributes need to be evaluated in the draw steps is an easy perf win

leebyp replied:
either in the drawer or drawsteps

hellochar replied:
Yep, eventually we should make the LineSVGDrawer put the "d" attribute, instead of the LinePlot creating the "d" attribute

themadcreator commented on src/drawers/rectangleDrawer.ts:
```diff
@@ -5,52 +5,54 @@
+    }, {} as { [key: string]: any | number | string });
+
+    context.beginPath();
+    context.rect(resolvedAttrs["x"], resolvedAttrs["y"], resolvedAttrs["width"], resolvedAttrs["height"]);
```
i believe in the scatter plot PR ben has pulled this out into a method that might be used by multiple canvas renders since many of them will want to set the styles from attrs.

leebyp replied:
Hehe I toyed with this exact idea of extending from separate `SVGDrawer`s and `CanvasDrawer`s when I was implementing the scatter plot (though not as fleshed out of course...) but wanted to defer till after getting the MVP out. If we're close here, I'd be happy to rebase.

themadcreator reviewed (commented):
Okay, this is pretty tremendous! I'm excited about this refactor but i have some thoughts, lemme know what you think

hellochar committed:
SVGDrawer now creates the _root

Drawer -> ProxyDrawer
introduce Drawer interface for SVGDrawer and CanvasDrawer

hellochar committed:
fix tests; upgrade to sinon2.1

hellochar committed:
migrate off deprecated sinon fn

hellochar commented:
@themadcreator I couldn't make the CanvasDrawer own creating/destroying the canvas element since we have one drawer per series, and 50 full-sized canvas elements is probably a bad idea. But I was able to build an interface for both of them which cleans stuff up, lemme know what you thnk!

leebyp commented on src/drawers/lineDrawer.ts:
```diff
@@ -5,56 +5,40 @@
+ * TODO put the d3.Line into the attrToAppliedProjector directly
+ */
+export function makeLineCanvasDrawStep(d3LineFactory: () => d3.Line<any>): CanvasDrawStep {
+  return (context: CanvasRenderingContext2D, data: any[][], attrToAppliedProjector: AttributeToAppliedProjector) => {
```
for consistency in getting at what's going on here, make this also a `const` function as `RectangleCanvasDrawStep`?

hellochar replied:
This is actually different from `RectangleCanvasDrawStep` though - that one is the same everywhere, but for line the draw logic changes depending on the d3LineFactory (although eventually we should remove that)

leebyp commented on src/drawers/canvasDrawer.ts:
```diff
@@ -0,0 +1,53 @@
+    // don't support animations for now; just draw the last draw step immediately
+    const lastDrawStep = appliedDrawSteps[appliedDrawSteps.length - 1];
+    this._context.save();
+    this._drawStep(this._context, data, lastDrawStep.attrToAppliedProjector);
```
do we still need the `setTimeout`? the svgDrawer is still using it

hellochar replied:
Funnily enough - `Window.Utils.setTimeout(..., 0)` just invokes the function directly, so this is exactly the same still. I'm not quite sure why we even need our own setTimeout

leebyp replied:
at the very least, we should remove from both or keep from both?

leebyp commented on src/drawers/rectangleDrawer.ts:
```diff
@@ -5,52 +5,49 @@
+import { SVGDrawer } from "./svgDrawer";
+
+export class RectangleSVGDrawer extends SVGDrawer {
+  constructor(private _rootClassName: string = null) {
```
`private _rootClassName: string = ""` will remove the need for the null checks?

hellochar committed:
save a null check by defaulting className to empty string

themadcreator commented on src/drawers/drawer.ts:
```diff
@@ -5,130 +5,96 @@
+/**
+ * Drawers draw data onto an output of some sort, usually a DOM element.
+ */
+export interface Drawer {
```
`Drawer` -> `IDrawer`

themadcreator commented on src/drawers/drawer.ts:
```diff
@@ -5,238 +5,101 @@
-  private _cachedSelection: SimpleSelection<any>;
-  private _cachedSelectionNodes: d3.BaseType[];
-
+export interface Drawer {
```
`Drawer` -> `IDrawer`

themadcreator commented on src/drawers/drawer.ts:
```diff
@@ -5,238 +5,101 @@
 
   /**
-   * Removes the Drawer and its renderArea
+   * Destroy this Drawer, remove all traces of it from the DOM, etc.
```
is this really true for canvas drawers?

themadcreator commented on src/drawers/drawer.ts:
```diff
@@ -5,238 +5,101 @@
-      Utils.Window.setTimeout(() => this._drawStepCanvas(data, lastDrawStep), 0);
-    } else {
-      throw new Error("Drawer's canvas and renderArea are both null!");
+  public useCanvas(canvas: d3.Selection<HTMLCanvasElement, any, any, any>) {
```
maybe it'd still be possible share one canvas across drawers without keeping track of it in the component, but this should be fine for now

themadcreator reviewed (approved):
Okay, i'm cool with this but rename the interface if you're inclined

hellochar committed:
Drawer -> IDrawer, improve comment

hellochar committed:
Merge branch 'develop' into xzhang/drawer-api-refactor

Conflicts:
	src/plots/barPlot.ts
	src/plots/commons.ts
	src/plots/piePlot.ts
	src/plots/plot.ts
	src/plots/segmentPlot.ts
	src/plots/waterfallPlot.ts
	test/drawers/drawerTests.ts
	test/plots/plotTests.ts
	test/scales/linearScaleTests.ts
	test/utils/transformAwareTranslatorTests.ts
